File: algorithms/1_easy/S/03903_smallest_stable_index_i/solution.c
int firstStableIndex(const int *restrict nums, const int len, const int k) {
    assert(len > 0 && nums != NULL && k >= 0);
    if (len == 1) { return 0; }

    int *restrict mins = malloc((size_t)len * sizeof(*mins));
    assert(mins != NULL);
    mins[len - 1] = nums[len - 1];
    for (int i = len - 2; i >= 0; i -= 1) {
        mins[i] = MIN(nums[i], mins[i + 1]);
    }

    int max = INT_MIN;
    int res = -1;
    for (int i = 0; i < len; i += 1) {
        max = MAX(max, nums[i]);
        if (max - mins[i] <= k) {
            res = i;
            break;
        }
    }

    free(mins);
    return res;
}

File: algorithms/1_easy/S/03903_smallest_stable_index_i/solution.rs
impl Solution {
    pub fn first_stable_index(nums: Vec<i32>, k: i32) -> i32 {
        assert!(!nums.is_empty() && k >= 0);
        if nums.len() == 1 { return 0; }

        let len = nums.len();
        let mut mins = nums.clone();
        for i in (0..(len - 1)).rev() {
            mins[i] = mins[i].min(mins[i + 1]);
        }

        let mut max = i32::MIN;
        for i in 0..len {
            max = max.max(nums[i]);
            if max - mins[i] <= k {
                return i as i32;
            }
        }

        -1
    }
}
